README.md
# Sigfox End-Point library implementation example on Holtek BC68F2150

## Description

This repository shows an example of Sigfox End-Point Library integraton on a HOLTEK low cost transmiter with flash MCU. It uses the code optimization features of the EP-LIB in order to minimize the memory footprint inside the MCU.
To have more informations about this design please refer to [wiki documentation](https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/wiki).

## Architecture

This repository contains several source code and project files needed to compile the Low Cost Device (LCD) project. Only the source code files of the EP_LIB are not be included inside this repository: the EP_LIB project is a submodule and must be updated before to upload sources files in `middleware/sigfox/sigfox-ep-lib` folder.

```
.                                # Root directory
├── middleware                   # Middleware
│   ├── sigfox                   # Sigfox middleware subdirectory
│   │    ├── sigfox-ep-lib       # EP_LIB submodule
│   │    │    └── ...
│   │    ├── manuf               # EP_LIB manufacturer API specific to this SoC
│   │    │    └── mcu_api.c
│   │    │    └── rf_api.c
│   │    │    └── mcu_rf_api_rodata.c
│   │    │    └── mcu_rf_api_rodata.h
│   │    │    └── TI_aes_128_encr_only.c
│   │    │    └── TI_aes_128_encr_only.h
├── driver                       # Drivers
│   │    ├── src                 # Driver header subdirectory
│   │    │    └── iap.h
│   │    ├── src                 # Driver source subdirectory
│   │    │    └── iap.c
├── output                       # Output generation folder, containing all generated files by IDE
│   ├── *.OBJ
│   ├── *.ASM
│   ├── *.LST
│   └── *...
├── main.c                       # Application source file
├── lcd.opt                      # Project option file
├── lcd.pjt                      # Project file
├── lcd.pjtx                     # Project file
└── README.md
```

## Hardware

The following figure shows the hardware architecture of the Holtek BC68F2150 board.

<p align="center">
<img src="https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/wiki/images/holtek_bc68f2150_architecture.drawio.png" width="600"/>
</p>

All gerbers and schematics are available in the wiki of this repository.

## Getting Started

### Cloning the repository and submodule

```bash
$ git clone https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150.git
$ cd sigfox-ep-example-holtek-bc68f2150
$ git submodule init
$ git submodule update
```

### How to build project

- Download the HT-IDE3000 software from Holtek web site.
- Click on pjtx project file to run IDE. Now you are ready to compile and update the source code for your application. *For more information on the IDE, compiler, etc... please refer to pdf files available in DOC IDE folder.*
- To configure the EP_LIB flags: right-click on the project > Project Settings > Project Option > Define Symbol.

### How to program and debug

- To flash or debug the device you need to buy a *e-link* programmer from HOLTEK
- Connect the device to the programmer and click on the Build/Rebuild button, the flash process will be launched automatically.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v1.3](https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/releases/tag/v1.3) - 22 Dec 2023

### Add

* Continous Wave implementation.

### Changed

* Change submodule address (ssh to https).
* Update **sigfox-ep-lib** submodule version from V3.3 to V3.4.
* Remove some comments.

## [v1.2](https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/releases/tag/v1.2) - 20 Oct 2023

### Changed

* Change **tx_power_dbm_eirp** to 14 dBm (maximum power).
* Change **number_of_frames** to 3 frame per message.
* Remove docs folder.

### Fixed

## [v1.1](https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/releases/tag/v1.1) - 16 Oct 2023

### Changed

* Improve **spectrum template** with new amplitude profile tables.
* Add **ERROR_CODES** return.

### Fixed

* Fix exection issues with new IDE version **HT-IDE3000 V8.1.5** .
* Fix All uplink library feature can be embedded on the Soc but it is strongly recommended to use the **optimization flags** according to targetted the application in order to limit the memory footprint.

### Known limitations

* **600 bps** not compliant with 1.7 EP-SPECS.

## [v1.0](https://github.com/sigfox-tech-radio/sigfox-ep-example-holtek-bc68f2150/releases/tag/v1.0) - 02 Jun 2023

### General

* First version of the Holtek BC68F2150 Sigfox example code.

### Known limitations

* Depending on the Holtek IDE version, some execution issues can appear on some specific conditions (full memory usage and/or call depth). Please consider using **HT-IDE3000 V8.1.4** or above.
* Due to the limited memory size (especially RAM) the full featured library cannot be embedded on the SoC. Use the **optimization flags** according to targetted the application.


main.c
/*!*****************************************************************
 * \file    main.c
 * \brief   Main program.
 *******************************************************************
 * \copyright
 *
 * Copyright (c) 2022, UnaBiz SAS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of UnaBiz SAS nor the names of its contributors may be
 *    used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
 * THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************/

#include "BC68F2150.h"
static const __attribute__((at(0x1D00))) unsigned char frame[12] = {0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA, 0xAA};
#include "sigfox_rc.h"
#include "sigfox_ep_api.h"

static void _delay_ms(unsigned short delay_ms) {
	unsigned short count_ms;
	for (count_ms=0 ; count_ms<delay_ms ; count_ms++) {
		GCC_DELAY(4000);	
	}
}


void main() {
	SIGFOX_EP_API_application_message_t SIGFOX_EP_API_application_message;
	SIGFOX_EP_API_config_t SIGFOX_EP_API_config; 
	//Watchdog
	_wdtc 	= 0b10101000;	//Disable WatchDog
	
	//Oscillator
	_rf_xo1 = 0b10000011;
	_hxtc	= 0b00000001;
	while(!_hxtf);
	_scc	= 0b00001000;	//Fh HXT
	
	//Inpout Output Port
	_pa_wake = 1;
	_pac 	= 0b11111111;
	_papu	= 0b11111111;
	_pawu   = 0b11111111;
	_pbc 	= 0b00000000;
	_pb 	= 0b00001111;
	
	_delay_ms(1000);
	_pb 	= 0b00000000;	
	SIGFOX_EP_API_config.rc = &SIGFOX_RC1;
	SIGFOX_EP_API_open(&SIGFOX_EP_API_config);
	SIGFOX_EP_API_application_message.common_parameters.t_ifu_ms = 500;
	SIGFOX_EP_API_application_message.common_parameters.tx_power_dbm_eirp = 14;
	SIGFOX_EP_API_application_message.type = SIGFOX_APPLICATION_MESSAGE_TYPE_BYTE_ARRAY;
	SIGFOX_EP_API_application_message.common_parameters.number_of_frames = 3;
	SIGFOX_EP_API_application_message.ul_payload_size_bytes = 12;
	SIGFOX_EP_API_application_message.ul_payload = frame;

	while(1) {
		// Trigger send with button.
		while (_pa1 != 0);
		_delay_ms(2000);
		SIGFOX_EP_API_open(&SIGFOX_EP_API_config);		
		SIGFOX_EP_API_send_application_message(&SIGFOX_EP_API_application_message);
		SIGFOX_EP_API_close();
	}
}


drivers/inc/iap.h
/*!*****************************************************************
 * \file    iap.h
 * \brief   In Application Programming driver.
 *******************************************************************
 * \copyright
 *
 * Copyright (c) 2022, UnaBiz SAS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of UnaBiz SAS nor the names of its contributors may be
 *    used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
 * THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************/

#ifndef __IAP_H__
#define __IAP_H__

/*** IAP structures ***/

/*!******************************************************************
 * \enum IAP_status_t
 * \brief In Application Programming error codes.
 *******************************************************************/
typedef enum {
	IAP_SUCCESS = 0,
	IAP_ERROR 	= 1,
} IAP_status_t;

/*!******************************************************************
 * \fn IAP_status_t IAP_ew_en_proc(void)
 * \brief In Application Programming erase write enable procedure.
 * \param[in]  	none
 * \param[out] 	none
 * \retval		Function execution status.
 *******************************************************************/
IAP_status_t IAP_ew_en_proc(void);

/*!******************************************************************
 * \fn IAP_status_t IAP_read(unsigned short add, unsigned short *flash_word, unsigned char len)
 * \brief In Application Programming read procedure.
 * \param[in]  	add:Address to read data
 * \param[out] 	*flash_word:Reading data pointer
 * \param[in] 	len:Size to read
 * \retval		Function execution status.
 *******************************************************************/
IAP_status_t IAP_read(unsigned short add, unsigned short *flash_word, unsigned char len);

/*!******************************************************************
 * \fn IAP_status_t IAP_write(unsigned short add, unsigned short *flash_word, unsigned char len, char erase_before)
 * \brief In Application Programming write procedure(warning the IAP drivers support only write data bloc of 4 word or 8 bytes).
 * \param[in]  	add:Address to write data
 * \param[out] 	*flash_word:Writing data pointer
 * \param[in] 	len:Size to write
 * \param[in] 	erase_before:Erase page or not before
 * \retval		Function execution status.
 *******************************************************************/
IAP_status_t IAP_write(unsigned short add, unsigned short *flash_word, unsigned char len, char erase_before);

#endif /* __IAP_H__ */


drivers/src/iap.c
/*!*****************************************************************
 * \file    iap.c
 * \brief   In Application Programming driver.
 *******************************************************************
 * \copyright
 *
 * Copyright (c) 2022, UnaBiz SAS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of UnaBiz SAS nor the names of its contributors may be
 *    used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
 * THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************/
 
#include "iap.h"
#include "BC68F2150.h"

IAP_status_t IAP_ew_en_proc(void) {
	_fmod0 = 0;
	_fmod1 = 1;
	_fmod2 = 1;
  	_fwpen = 1;
	_fd1l = 0x00;
  	_fd1h = 0x04;
  	_fd2l = 0x0d;
  	_fd2h = 0x09;
  	_fd3l = 0xc3;
  	_fd3h = 0x40;
  	while (_fwpen) 
  		GCC_CLRWDT();
  	if (_cfwen == 0)
  		return IAP_ERROR;
  	else
  		return IAP_SUCCESS;
}
IAP_status_t IAP_read(unsigned short add, unsigned short *flash_word, unsigned char len) {
	//Configure to read program memory and set flash memory read enable bit
	_fc0 &= 0b10001101;
	_fc0 |= 0b00110010;
  	while(len != 0) {
  		//Set address to read
  		_farh = (add >> 8) & 0x1F;
  		_farl = (add) & 0xFF;
  		_frd = 1;
  		//Wait read process completed 
  		while (_frd)
    		GCC_CLRWDT();
    	//Read lower byte	
  		*flash_word = (_fd0l & 0x00FF);
  		//Read upper byte
 		*flash_word |= ((_fd0h << 8) & 0xFF00);
 		flash_word++;
 		len --;
  		//Incr address to next 2 or 1 bytes	
  		add += 1;
  	}; 
  	//Clear read enable bit
  	_frden = 0;
	return IAP_SUCCESS;
}

IAP_status_t IAP_write(unsigned short add, unsigned short *flash_word, unsigned char len, char erase_before) {
	
	if (IAP_ew_en_proc() != IAP_SUCCESS)
		return IAP_ERROR;
	
	if (erase_before) {
  	//Erase Block
  		_farh = (add >> 8) & 0x1F;
  		_farl = (add) & 0xFF;
		_fmod0 = 1;
		_fmod1 = 0;
		_fmod2 = 0;
		_fwt = 1;
	    while (_fwt != 0);	    
	}
	
	while(len != 0) {
		_fmod0 = 0;
    	_fmod1 = 0;
   	 	_fmod2 = 0;
   	 	_fd0l = 0xFF;
	  	_fd0h = 0xFF;
		_fd1l = 0xFF;
	  	_fd1h = 0xFF;
	  	_fd2l = 0xFF;
	  	_fd2h = 0xFF;
	  	_fd3l = 0xFF;
	  	_fd3h = 0xFF;

		_fd0l = (*flash_word) & 0xFF;
		_fd0h = (*flash_word >> 8) &  0xFF;
		flash_word++;
		len--;
		if (len == 0) {
			_fwt = 1;
			while (_fwt != 0);	
			continue;
		}
		_fd1l = (*flash_word) & 0xFF;
		_fd1h = (*flash_word >> 8) &  0xFF;
		flash_word++;
		len--;
		if (len == 0) {
			_fwt = 1;
			while (_fwt != 0);	
			continue;
		}
		_fd2l = (*flash_word) & 0xFF;
		_fd2h = (*flash_word >> 8) &  0xFF;
		flash_word++;
		len--;
		if (len == 0) {
			_fwt = 1;
			while (_fwt != 0);	
			continue;
		}
		_fd3l = (*flash_word) & 0xFF;
		_fd3h = (*flash_word >> 8) &  0xFF;
		flash_word++;
		len--;
		_fwt = 1;
		while (_fwt != 0);
	}
	_cfwen = 0;
	return IAP_SUCCESS;
}


middleware/sigfox/manuf/mcu_rf_api_rodata.c
/*!*****************************************************************
 * \file    mcu_rf_rodata.c
 * \brief   rodata file.
 *******************************************************************
 * \copyright
 *
 * Copyright (c) 2022, UnaBiz SAS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of UnaBiz SAS nor the names of its contributors may be
 *    used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
 * THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************/

#include "mcu_rf_api_rodata.h"

const MCU_RF_RODATA_t __attribute__((at(0x1E00)))  mcu_rf_rodata = {
	.id 		= {	0xFE, 0xDC, 0xBA,  0x98},
	.key 		= {	0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF, 0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF},
	.pac	 	= {	0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00},
	.pwrtab_100bps_HP = {	.size = 100,
							.tab= {	0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,
									0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,
									0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,0x26,
									0x26,0x26,0x26,0x26,0x26,0x26,0x26,
									0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,0x66,
									0xA6,0xA6,0xA6,0xA6,0xA6,0xA6,0xA6,0xA6,0xA6,
									0xE6,0xE6,0xE6,0xE6,0xE6,0xE6,
									0x56,0x56,0x56,0x56,0x56,0x56,
									0x54,0x54,0x54,0x54,0x54,
					                0x94,0x94,0x94,
					                0xD4,0xD4,0xD4,
					                0xD2,0xD2,0xD2,0xD2,
					                0x04,0x04,0x04,0x04,0x04,
									0x02,0x02,0x02,0x02,
									0xC0,0xC0,
									},
							.delta_f = 656,
						},
	.pwrtab_100bps_LP = {	.size = 100,
							.tab = {0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,
									0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,
									0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,
									0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,0x24,
									0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,0x64,
									0xA4,0xA4,0xA4,0xA4,0xA4,0xA4,0xA4,
									0xE4,0xE4,0xE4,0xE4,
									0x14,0x14,0x14,0x14,
									0x54,0x54,0x54,0x54,
						        	0x94,0x94,0x94,0x94,
						            0x92,0x92,0x92,0x92,
						            0xD2,0xD2,0xD2,0xD2,
						            0x04,0x04,0x04,0x04,0x04,0x04,0x04,
									0x02,0x02,0x02,0x02,
									0xC0,0xC0,
							},
							.delta_f = 656,	
						},
	.pwrtab_600bps_LP	= {0x00},				
	.pwrtab_600bps_LP = {	.size = 22,
							.tab = {0x24,0x24,0x24,0x24,0x24,0x24,
								0x64,0x64,0x64,0x64,
								0xA4,0xA4,0xA4,
								0xE4,
								0x14,
								0x54,
					        	0x94,
					            0xD4,
					            0xD2,
					            0x04,
								0x02,
								0xC0,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
								0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
							},
							.delta_f = 866,	
						},						
	
	.version	= "V1.0",
	.empty		= {0x00},			
};

const unsigned short __attribute__((at(0x1F00))) nvm[256]= {0x0000};

middleware/sigfox/manuf/mcu_api.c
/*!*****************************************************************
 * \file    rf_api.c
 * \brief   Radio drivers.
 *******************************************************************
 * \copyright
 *
 * Copyright (c) 2022, UnaBiz SAS
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 *  1 Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 *  2 Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  3 Neither the name of UnaBiz SAS nor the names of its contributors may be
 *    used to endorse or promote products derived from this software without
 *    specific prior written permission.
 *
 * NO EXPRESS OR IMPLIED LICENSES TO ANY PARTY'S PATENT RIGHTS ARE GRANTED BY
 * THIS LICENSE. THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
 * CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************/

#include "manuf/mcu_api.h"
#include "sigfox_types.h"
#include "sigfox_error.h"
#include "mcu_rf_api_rodata.h"
#include "TI_aes_128_encr_only.h"
#include "BC68F2150.h"
#include "iap.h"

#ifdef TIMER_REQUIRED
MCU_API_status_t MCU_API_timer_start(MCU_API_timer_t *timer) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif
	unsigned short i;
	
	for(i = 0; i < timer->duration_ms; i++) {
		GCC_DELAY(4000);
	}
	
	RETURN();
}
#endif

#ifdef TIMER_REQUIRED
MCU_API_status_t MCU_API_timer_stop(MCU_API_timer_instance_t timer_instance) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif
	
	RETURN();
}
#endif

#if (defined TIMER_REQUIRED) && !(defined ASYNCHRONOUS)
MCU_API_status_t MCU_API_timer_status(MCU_API_timer_instance_t timer_instance, sfx_bool *timer_has_elapsed) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif

	RETURN();
}
#endif

#if (defined TIMER_REQUIRED) && !(defined ASYNCHRONOUS)
MCU_API_status_t MCU_API_timer_wait_cplt(MCU_API_timer_instance_t timer_instance) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif

	RETURN();
}
#endif

MCU_API_status_t MCU_API_aes_128_cbc_encrypt(MCU_API_encryption_data_t *aes_data) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif

#ifdef PUBLIC_KEY_CAPABLE
    if (aes_data->key == SIGFOX_EP_KEY_PRIVATE) {
        aes_encrypt(aes_data->data, (sfx_u8 *)mcu_rf_rodata.key);
    } else if ((aes_data->key == SIGFOX_EP_KEY_PUBLIC)) {
        aes_encrypt(aes_data->data, (sfx_u8 *)SIGFOX_EP_PUBLIC_KEY);
    }
#else
    aes_encrypt(aes_data->data, (sfx_u8 *) mcu_rf_rodata.key);
#endif

	RETURN();
}

MCU_API_status_t MCU_API_get_ep_id(sfx_u8 *ep_id, sfx_u8 ep_id_size_bytes) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif
	
	IAP_read(0x1E00, (unsigned short*)ep_id, ep_id_size_bytes / 2);

	RETURN();
}

MCU_API_status_t MCU_API_get_nvm(sfx_u8 *nvm_data, sfx_u8 nvm_data_size_bytes) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif
	IAP_read(0x1F00, (unsigned short*)nvm_data, nvm_data_size_bytes / 2);
	
	RETURN();
}

MCU_API_status_t MCU_API_set_nvm(sfx_u8 *nvm_data, sfx_u8 nvm_data_size_bytes) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif
	IAP_write(0x1F00, (unsigned short*)nvm_data, nvm_data_size_bytes / 2, 1);
	
	RETURN();
}

#ifdef VERBOSE
MCU_API_status_t MCU_API_get_initial_pac(sfx_u8 *initial_pac, sfx_u8 initial_pac_size_bytes) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif

	RETURN();
}
#endif

#ifdef VERBOSE
MCU_API_status_t MCU_API_get_version(sfx_u8 **version, sfx_u8 *version_size_char) {
#ifdef ERROR_CODES
    MCU_API_status_t status = MCU_API_SUCCESS;
#endif

	/* To be implemented by the device manufacturer */
    (*version) = (sfx_u8*) mcu_rf_rodata.version;
    (*version_size_char) = (sfx_u8) sizeof(mcu_rf_rodata.version);
    
    RETURN();
}
#endif

#ifdef ERROR_CODES
/*******************************************************************/
void MCU_API_error(void) {
	/* To be implemented by the device manufacturer */
}
#endif